File: 22.generate-parentheses.go
/*
 * @lc app=leetcode id=22 lang=golang
 *
 * [22] Generate Parentheses
 *
 * https://leetcode.com/problems/generate-parentheses/description/
 *
 * algorithms
 * Medium (53.96%)
 * Total Accepted:    320.7K
 * Total Submissions: 592.9K
 * Testcase Example:  '3'
 *
 * 
 * Given n pairs of parentheses, write a function to generate all combinations
 * of well-formed parentheses.
 * 
 * 
 * 
 * For example, given n = 3, a solution set is:
 * 
 * 
 * [
 * ⁠ "((()))",
 * ⁠ "(()())",
 * ⁠ "(())()",
 * ⁠ "()(())",
 * ⁠ "()()()"
 * ]
 * 
 */

 func isValid(s string) bool {
	if s==""{
		return true
	}
	stackS:=[]byte{}
	for i:=0;i<len(s);i++{
		if s[i]=='{' ||s[i]=='['||s[i]=='('{
			stackS=append(stackS,s[i])
		}else{
			if len(stackS)==0{
				return false
			}
			if (stackS[len(stackS)-1]=='{' && s[i]=='}')||(stackS[len(stackS)-1]=='[' &&s[i]==']')||(stackS[len(stackS)-1]=='(' && s[i]==')'){
				stackS = stackS[0:len(stackS)-1]	
			}else{
				return false
			}
		}
	}
	if len(stackS)==0{
		return true
	}
	return false
}

func generateParenthesis(n int) []string {
	paren:=""
}

File: 22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#
from typing import *
import copy
# @lc code=start
class Solution:
    def __init__(self):
        self.left = 0
        self.right = 0
        self.parentheses = []
        self.parenthese = []
    def generateParenthesis2(self, n: int) -> List[str]:
        n = n * 2
        self.helper(n,0)  
        return self.parentheses
        
    def helper(self,n,k):
        if self.right > self.left:
            return 
        if n==k:
            if self.left == self.right :
                self.parentheses.append("".join(self.parenthese))
            return
        self.parenthese.append("(")
        self.left += 1
        self.helper(n,k+1)
        self.left -= 1
        self.parenthese.pop()

        self.parenthese.append(")")
        self.right += 1
        self.helper(n,k+1)
        self.right -= 1
        self.parenthese.pop()

    def generateParenthesis(self, n: int) -> List[str]:
        parenthesis = [''] * (n * 2)
        left = 0
        right = 0
        
        for s in range(['(',')']):
            for i in range(n*2):

                        
# @lc code=end

if __name__ == "__main__":
    print(Solution().generateParenthesis(3))